Next batch of ngtsc changes

alxhub opened the pull request.

alxhub committed:
fix(ivy): handle ReadKeyExpr code generation

This implements a missing expression type in ngtsc code generation:
that of bracket access to an object property.

alxhub committed:
fix(ivy): allow building router with ngtsc

This commit adds the ivy-local tag to //packages/router. Since the
router depends on //packages/upgrade, it makes that package
compatible with ngtsc as well.

alxhub committed:
fix(ivy): unwrap parenthesized or cast expressions for metadata

Metadata in Ivy must be literal. For example,

@NgModule({...})

is legal, whereas

const meta = {...};
@NgModule(meta)

is not.

However, some code contains additional superfluous parentheses:

@NgModule(({...}))

It is desirable that ngtsc accept this form of literal object.

alxhub committed:
feat(ivy): selector side of ModuleWithProviders via type metadata

Within an @NgModule it's common to include in the imports a call to
a ModuleWithProviders function, for example RouterModule.forRoot().
The old ngc compiler was able to handle this pattern because it had
global knowledge of metadata of not only the input compilation unit
but also all dependencies.

The ngtsc compiler for Ivy doesn't have this knowledge, so the
pattern of ModuleWithProviders functions is more difficult. ngtsc
must be able to determine which module is imported via the function
in order to expand the selector scope and properly tree-shake
directives and pipes.

This commit implements a solution to this problem, by adding a type
parameter to ModuleWithProviders through which the actual module
type can be passed between compilation units.

The provider side isn't a problem because the imports are always
copied directly to the ngInjectorDef.

alxhub committed:
refactor(ivy): fix an unnecessarily deep import

alxhub committed:
feat(ivy): statically resolve template expressions

This commit adds support for template substitution expressions for
ngtsc static resolution.

alxhub committed:
fix(ivy): export InheritDefinitionFeature

alxhub committed:
fix(ivy): properly inject all special token types

Previously ngtsc had a few bugs handling special token types:

* Injector was not properly translated to INJECTOR
* ChangeDetectorRef was not injected via injectChangeDetectorRef()

This commit fixes these two bugs, and also adds a test to ensure
they continue to work correctly.

alxhub committed:
fix(ivy): ngInjectorDef should copy full imports/exports nodes

@NgModule()s get compiled to two fields: ngModuleDef and ngInjectorDef.
Both fields contain imports, as both selector scopes and injectors have
the concept of composed units of configuration. Previously these fields
were generated by static resolution of imports and exports in metadata.

Support for ModuleWithProviders requires they be generated differently.
ngModuleDef's imports/exports are generated as resolved lists of types,
whereas ngInjectorDef's imports should reflect the raw expressions that
the developer wrote in the metadata.

This change modifies the NgModule handler and properly copies raw nodes
for the imports and exports into the ngInjectorDef.

alxhub committed:
fix(ivy): export NgModuleFactory adapter

alxhub committed:
fix(ivy): ignore imports without ngInjectorDef in r3_injector

ngInjectorDef.imports is generated from @NgModule.imports plus
@NgModule.exports. A problem arises as a result, because @NgModule
exports contain not only other modules (which will have ngInjectorDef
fields), but components, directives, and pipes as well. Because of
locality, it's difficult for the compiler to filter these out at
build time.

It's not impossible, but for now filtering them out at runtime will
allow testing of the compiler against complex applications.

alxhub committed:
fix(ivy): fix a couple issues with Input/Output compilation

alxhub committed:
feat(ivy): port the static resolver to use the ReflectionHost

Previously, the static resolver did its own interpretation of statements
in the TypeScript AST, which only functioned on TypeScript code. ES5
code in particular would not work with the resolver as it had hard-coded
assumptions about AST structure.

This commit changes the resolver to use a ReflectionHost instead, which
abstracts away understanding of the structural side of the AST. It adds 3
new methods to the ReflectionHost in support of this functionality:

* getDeclarationOfIdentifier
* getExportsOfModule
* isClass

alxhub committed:
fix(ivy): detect ngOnChanges as a non-static method

Previously ngtsc had a bug where it would only detect the presence of
ngOnChanges as a static method. This commit flips the condition and only
recognizes ngOnChanges as a non-static method.

alxhub committed:
fix(ivy): references track the identifier they were discovered under

Previously, references had the concept of an identifier, but would not
properly detect whether the identifier should be used or not when
generating an expression. This change fixes that logic.

Additionally, now whenever an identifier resolves to a reference (even
one imported from another module) as part of resolving an expression,
the reference is updated to use that identifier. This ensures that for
a class Foo declared in foo.ts, but referenced in an expression in
bar.ts, the Reference returned includes the identifier from bar.ts,
meaning that writing an expression in bar.ts for the Reference will not
generate an import.

alxhub committed:
fix(ivy): wrap non-statement assignment expressions in parentheses

Previously, when translating an assignment expression (e.g. x = 3), the
translator would always print the statement as X = Y. However, if the
expression is included in a larger expression (X = (Y = Z)), the
translator would print "X = Y = Z" without regard for the outer
expression context.

Now, the translator understands when it's printing an expression
statement (X = Y;) vs an expression in a larger context (X = (Y = Z);)
and encapsulates the latter in parentheses.

alxhub committed:
refactor(ivy): run the compiler compliance tests against ngtsc

This commit moves the compiler compliance tests into compiler-cli,
and uses ngtsc to run them instead of the custom compilation
pipeline used before. Testing against ngtsc allows for validation
of the real compiler output.

This commit also fixes a few small issues that prevented the tests
from passing.

mary-poppins commented:
You can preview 0e78341 at https://pr24862-0e78341.ngbuilds.io/.

mary-poppins commented:
You can preview 3b182aa at https://pr24862-3b182aa.ngbuilds.io/.

mary-poppins commented:
You can preview 3dd7c1d at https://pr24862-3dd7c1d.ngbuilds.io/.

alxhub committed:
fix(ivy): correctly export all *Def symbols as private

Previously, some of the *Def symbols were not exported or were exported
as public API. This commit ensures every definition type is in the
private export namespace.

alxhub committed:
fix(ivy): export injectElementRef

alxhub committed:
fix(ivy): types in .d.ts files should account for generics

Ivy definition types have a generic type which specifies the return
type of the factory function. For example:

static ngDirectiveDef<NgForOf, '[ngFor][ngForOf]'>

However, in this case NgForOf itself has a type parameter <T>. Thus,
writing the above is incorrect.

This commit modifies ngtsc to understand the genericness of NgForOf and
to write the following:

static ngDirectiveDef<NgForOf<any>, '[ngFor][ngForOf]'>

mary-poppins commented:
You can preview 6fcbc8f at https://pr24862-6fcbc8f.ngbuilds.io/.

mary-poppins commented:
You can preview 9e3851c at https://pr24862-9e3851c.ngbuilds.io/.

mary-poppins commented:
You can preview e34dec4 at https://pr24862-e34dec4.ngbuilds.io/.

alxhub committed:
fix(ivy): use 'typeof' and 'never' for type metadata

Previously ngtsc would use a tuple of class types for listing metadata
in .d.ts files. For example, an @NgModule's declarations might be
represented with the type:

[NgIf, NgForOf, NgClass]

If the module had no declarations, an empty tuple [] would be produced.

This has two problems.

1. If the class type has generic type parameters, TypeScript will
complain that they're not provided.

2. The empty tuple type is not actually legal.

This commit addresses both problems.

1. Class types are now represented using the `typeof` operator, so the
above declarations would be represented as:

[typeof NgIf, typeof NgForOf, typeof NgClass].

Since typeof operates on a value, it doesn't require generic type
arguments.

2. Instead of an empty tuple, `never` is used to indicate no metadata.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -87,3 +87,10 @@ export function referenceToExpression(ref: Reference, context: ts.SourceFile): E
   return decorator.import !== null && decorator.import.from === '@angular/core';
 }
+
+export function unwrapExpression(node: ts.Expression): ts.Expression {
```
add a one liner comment ?

alxhub replied:
Done.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -125,6 +127,46 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
+   * Given a `FunctionDeclaration` or `MethodDeclaration`, check if it is typed as a
+   * `ModuleWithProviders` and return an expression referencing the module if available.
+   */
+  private _resolveModuleWithProviders(node: ts.FunctionDeclaration|
```
IMO the name doesn't reflect what the method does

alxhub replied:
Renamed to `_extractModuleFromModuleWithProvidersFn`

vicb commented on packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts:
```diff
@@ -179,9 +181,14 @@ export class AbsoluteReference extends Reference {
-      node, {absoluteModuleName: null, scope: new Map<ts.ParameterDeclaration, ResolvedValue>()});
+export function staticallyResolve(
+    node: ts.Expression, checker: ts.TypeChecker,
+    foreignFunctionResolver?: (node: ts.FunctionDeclaration | ts.MethodDeclaration) =>
```
pls update the api doc

alxhub replied:
Done.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
+    predicate = new WrappedNodeExpr(args[0]);
+  } else if (typeof arg === 'string') {
+    predicate = [arg];
+  } else if (assertIsStringArray(arg, '@' + name)) {
```
`assert` are often debug time checks only, rename/refactor ?

alxhub replied:
Renamed to `isStringArrayOrDie()`

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
+  } else if (assertIsStringArray(arg, '@' + name)) {
+    predicate = arg as string[];
+  }
+  if (predicate === null) {
```
should it be a final `else {}` ?

alxhub replied:
Done.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
+    }
+
+    if (options.has('descendants')) {
+      const descendantsValue = staticallyResolve(options.get('descendants') !, checker);
```
nit: remove the local variable and use `descendants` ?

alxhub replied:
The assignment isn't legal until after the `typeof` check.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
+
+  // Extract the read and descendants options.
+  let read: Expression|null = null;
+  let descendants: boolean = true;
```
shouldn't `descendants` be a function of `first` ? (I think the behavior is different, can't remember exactly)

alxhub replied:
Fixed.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -199,3 +302,109 @@ function parseDecoratedFields(
+            if (decorator.args.length === 2) {
+              const resolvedArgs = staticallyResolve(decorator.args[1], checker);
+              if (!assertIsStringArray(resolvedArgs, '@HostListener.args')) {
+                throw new Error(`@HostListener arguments must be a string array`);
```
nit: `arguments` -> `second argument` would be clearer IMO

alxhub replied:
Fixed.

vicb commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -412,4 +412,64 @@ describe('ngtsc behavioral tests', () => {
+        })
+        class FooCmp {
+          @ContentChild('bar', {read: TemplateRef}) child: any;
+          @ContentChildren(TemplateRef) children: any;
```
Add tests for `@View...` and getters / setters ?

vicb commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -412,4 +412,64 @@ describe('ngtsc behavioral tests', () => {
+          template: 'Test',
+          host: {
+            '[attr.hello]': 'foo',
+            '(click)': 'onClick($event)',
```
add test for args ?

vicb commented on packages/compiler-cli/src/ngtsc/transform/src/translator.ts:
```diff
@@ -6,10 +6,19 @@
+export class Context {
+  constructor(readonly isStatement: boolean) {}
+
+  get setExpressionMode(): Context { return this.isStatement ? new Context(false) : this; }
```
nit: weird that a getter as side effect. make it a regular method ?

vicb reviewed (approved):
Only a few minor comments, mostly need.

Seeing the size of this PR frightened me at first but HUGE kudos to you for:
- Splitting the PR in small and scoped commits,
- Adding exhaustive PR comments to help understanding where it makes sense,
- Adding inline docs were it makes sense.

It turned out to be a very pleasant PR to review - thanks for easing the review.

mary-poppins commented:
You can preview ffec1f1 at https://pr24862-ffec1f1.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts:
```diff
@@ -356,6 +358,22 @@ class StaticInterpreter {
 
+  private visitTemplateExpression(node: ts.TemplateExpression, context: Context): ResolvedValue {
+    const pieces: string[] = [node.head.text];
+    for (let i = 0; i < node.templateSpans.length; i++) {
```
Tagged templates are not handled yet, so they'll fall through and return `DYNAMIC_VALUE`. I don't even know if `ngc` handles them.

alxhub committed:
feat(ivy): compile queries in ngtsc

This commit adds support for @ContentChild[ren] and @ViewChild[ren] in
ngtsc. Previously queries were ignored.

alxhub committed:
feat(ivy): support host bindings in ngtsc

This change adds support for host bindings to ngtsc, and parses them
both from decorators and from the metadata in the top-level annotation.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -80,17 +80,19 @@ export function extractDirectiveMetadata(
   const decoratedElements =
       members.filter(member => !member.isStatic && member.decorators !== null);
 
+  const coreModule = isCore && '@angular/core' || undefined;
```
Done.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -109,15 +109,29 @@ export class ComponentDecoratorHandler implements DecoratorHandler<R3ComponentMe
+    }
+
+    // Construct the list of view queries.
+    const coreModule = this.isCore && '@angular/core' || undefined;
```
Done.

mary-poppins commented:
You can preview 7b8acbc at https://pr24862-7b8acbc.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -109,15 +112,31 @@ export class ComponentDecoratorHandler implements DecoratorHandler<R3ComponentMe
+    }
+
+    // Construct the list of view queries.
+    const coreModule = this.isCore ? '@angular/core' : undefined;
```
Hmm, during core compilation, `coreModule` should be undefined right? This seems inverted now.

mary-poppins commented:
You can preview 81bff87 at https://pr24862-81bff87.ngbuilds.io/.

mary-poppins commented:
You can preview ee3ecf7 at https://pr24862-ee3ecf7.ngbuilds.io/.
